pkg/querier/batch/batch.go

```diff
@@ -83,7 +83,13 @@ func newIteratorAdapter(underlying iterator) chunkenc.Iterator {
 
 // Seek implements storage.SeriesIterator.
 func (a *iteratorAdapter) Seek(t int64) bool {
-	// Optimisation: see if the seek is within the current batch.
+	// Optimisations: see if the seek is before or within the current batch.
+	if a.curr.Length > 0 && a.curr.Index < a.curr.Length && t <= a.curr.Timestamps[0] {
```

Comment:
This comment does not match the code: the test does not look at whether the seek is "within" the batch, only whether it is before.

I think it would be more consistent to check `t <= a.curr.Timestamps[a.curr.Index]`, then the comment could be "before the current sample", which matches the interface comment `// If current sample found by previous ...`

Reply:
thanks for the comment @bboreham.  The `// Optimisations: see if the seek is before or within the current batch.` comment is actually meant for both the new and old if statements; namely: 

`if a.curr.Length > 0 && t >= a.curr.Timestamps[0] && t <= a.curr.Timestamps[a.curr.Length-1]` 

and 

`if a.curr.Length > 0 && a.curr.Index < a.curr.Length && t <= a.curr.Timestamps[0]`. 

Because the the 2 if statements combined are the optimizations. Notice how I changed `Optimisation` to `OptimisationS` :).  

But yes, I can see how the comment is confusing, I will think of a way to make it more clear. May be break the comment into two. I like the `t <= a.curr.Timestamps[a.curr.Index]` idea, makes the code more self explanatory too.

Reply:
> Also, I believe the result of discussions is that the code (unchanged by this PR) does not match the interface description in some cases: can you raise issues for those?

For sure, I have that as to do on my sticky notes.

Reply:
> comment is actually meant for both the new and old if statements

If you are trying to describe these two things together then you should include an explanation why the behaviour is different between the two - `a.curr.Index` is reset in one case and not the other.

(which, if you change to `t <= a.curr.Timestamps[a.curr.Index]`, will probably be easier)

Reply:
> If you are trying to describe these two things together then you should include an explanation why the behaviour is different between the two - `a.curr.Index` is reset in one case and not the other.

Yea, I actually have no idea why `a.curr.Index` was reset in the original code. 
 
> (which, if you change to `t <= a.curr.Timestamps[a.curr.Index]`, will probably be easier)

Yup I made this change along with some other changes in attempt to make the code more clear. I also take out the reset of `a.curr.Index`. Have a look and let me know if it make sense to you. But effectively with the `t <= a.curr.Timestamps[a.curr.Index]` we fix one of the but where the iterator can go backward; hopefully no existing code depend on the ability of batch iterator being able to go backward; I have been looking around (non-exhaustively), and I didn't find any code that depends on the go-backward-capability.

Reply:
Initially I thought seeking backward should be allowed, but the description on the interface is `Seek advances the iterator forward ...`, so it seems fair enough to interpret as "seeking backward not allowed".

Maybe we could ask to clarify the comment on the interface in Prometheus.